File: frontend/v1/src/models/validIngredient.ts
import * as Factory from "factory.ts";
import faker from "faker";
import {defaultFactories} from "@/models/fakes";

export class ValidIngredient {
  id: number;
  name: string;
  variant: string;
  description: string;
  warning: string;
  containsEgg: boolean;
  containsDairy: boolean;
  containsPeanut: boolean;
  containsTreeNut: boolean;
  containsSoy: boolean;
  containsWheat: boolean;
  containsShellfish: boolean;
  containsSesame: boolean;
  containsFish: boolean;
  containsGluten: boolean;
  animalFlesh: boolean;
  animalDerived: boolean;
  measurableByVolume: boolean;
  icon: string;
  createdOn: number;
  updatedOn?: number;
  archivedOn?: number;

  constructor() {
    this.id = 0;
    this.name = "";
    this.variant = "";
    this.description = "";
    this.warning = "";
    this.containsEgg = false;
    this.containsDairy = false;
    this.containsPeanut = false;
    this.containsTreeNut = false;
    this.containsSoy = false;
    this.containsWheat = false;
    this.containsShellfish = false;
    this.containsSesame = false;
    this.containsFish = false;
    this.containsGluten = false;
    this.animalFlesh = false;
    this.animalDerived = false;
    this.measurableByVolume = false;
    this.icon = "";
    this.createdOn = 0;
  }

static areEqual = function(
  vi1: ValidIngredient,
  vi2: ValidIngredient,
): boolean {
    return (
      vi1.id === vi2.id &&
      vi1.name === vi2.name &&
      vi1.variant === vi2.variant &&
      vi1.description === vi2.description &&
      vi1.warning === vi2.warning &&
      vi1.containsEgg === vi2.containsEgg &&
      vi1.containsDairy === vi2.containsDairy &&
      vi1.containsPeanut === vi2.containsPeanut &&
      vi1.containsTreeNut === vi2.containsTreeNut &&
      vi1.containsSoy === vi2.containsSoy &&
      vi1.containsWheat === vi2.containsWheat &&
      vi1.containsShellfish === vi2.containsShellfish &&
      vi1.containsSesame === vi2.containsSesame &&
      vi1.containsFish === vi2.containsFish &&
      vi1.containsGluten === vi2.containsGluten &&
      vi1.animalFlesh === vi2.animalFlesh &&
      vi1.animalDerived === vi2.animalDerived &&
      vi1.measurableByVolume === vi2.measurableByVolume &&
      vi1.icon === vi2.icon &&
      vi1.archivedOn === vi2.archivedOn
    );
  }
}

export const fakeValidIngredientFactory = Factory.Sync.makeFactory<ValidIngredient> ({
  name: Factory.Sync.each(() =>  faker.random.word()),
  variant: Factory.Sync.each(() =>  faker.random.word()),
  description: Factory.Sync.each(() =>  faker.random.word()),
  warning: Factory.Sync.each(() =>  faker.random.word()),
  containsEgg: Factory.Sync.each(() =>  faker.random.boolean()),
  containsDairy: Factory.Sync.each(() =>  faker.random.boolean()),
  containsPeanut: Factory.Sync.each(() =>  faker.random.boolean()),
  containsTreeNut: Factory.Sync.each(() =>  faker.random.boolean()),
  containsSoy: Factory.Sync.each(() =>  faker.random.boolean()),
  containsWheat: Factory.Sync.each(() =>  faker.random.boolean()),
  containsShellfish: Factory.Sync.each(() =>  faker.random.boolean()),
  containsSesame: Factory.Sync.each(() =>  faker.random.boolean()),
  containsFish: Factory.Sync.each(() =>  faker.random.boolean()),
  containsGluten: Factory.Sync.each(() =>  faker.random.boolean()),
  animalFlesh: Factory.Sync.each(() =>  faker.random.boolean()),
  animalDerived: Factory.Sync.each(() =>  faker.random.boolean()),
  measurableByVolume: Factory.Sync.each(() =>  faker.random.boolean()),
  icon: Factory.Sync.each(() =>  faker.random.word()),
  ...defaultFactories,
});

File: models/v1/valid_ingredient.go
package models

import (
	"context"
	"net/http"

	"gitlab.com/prixfixe/prixfixe/internal/v1/search"
)

const (
	// ValidIngredientsSearchIndexName is the name of the index used to search through valid ingredients.
	ValidIngredientsSearchIndexName search.IndexName = "valid_ingredients"
)

type (
	// ValidIngredient represents a valid ingredient.
	ValidIngredient struct {
		ID                 uint64  `json:"id"`
		Name               string  `json:"name"`
		Variant            string  `json:"variant"`
		Description        string  `json:"description"`
		Warning            string  `json:"warning"`
		ContainsEgg        bool    `json:"containsEgg"`
		ContainsDairy      bool    `json:"containsDairy"`
		ContainsPeanut     bool    `json:"containsPeanut"`
		ContainsTreeNut    bool    `json:"containsTreeNut"`
		ContainsSoy        bool    `json:"containsSoy"`
		ContainsWheat      bool    `json:"containsWheat"`
		ContainsShellfish  bool    `json:"containsShellfish"`
		ContainsSesame     bool    `json:"containsSesame"`
		ContainsFish       bool    `json:"containsFish"`
		ContainsGluten     bool    `json:"containsGluten"`
		AnimalFlesh        bool    `json:"animalFlesh"`
		AnimalDerived      bool    `json:"animalDerived"`
		MeasurableByVolume bool    `json:"measurableByVolume"`
		Icon               string  `json:"icon"`
		CreatedOn          uint64  `json:"createdOn"`
		LastUpdatedOn      *uint64 `json:"lastUpdatedOn"`
		ArchivedOn         *uint64 `json:"archivedOn"`
	}

	// ValidIngredientList represents a list of valid ingredients.
	ValidIngredientList struct {
		Pagination
		ValidIngredients []ValidIngredient `json:"validIngredients"`
	}

	// ValidIngredientCreationInput represents what a user could set as input for creating valid ingredients.
	ValidIngredientCreationInput struct {
		Name               string `json:"name"`
		Variant            string `json:"variant"`
		Description        string `json:"description"`
		Warning            string `json:"warning"`
		ContainsEgg        bool   `json:"containsEgg"`
		ContainsDairy      bool   `json:"containsDairy"`
		ContainsPeanut     bool   `json:"containsPeanut"`
		ContainsTreeNut    bool   `json:"containsTreeNut"`
		ContainsSoy        bool   `json:"containsSoy"`
		ContainsWheat      bool   `json:"containsWheat"`
		ContainsShellfish  bool   `json:"containsShellfish"`
		ContainsSesame     bool   `json:"containsSesame"`
		ContainsFish       bool   `json:"containsFish"`
		ContainsGluten     bool   `json:"containsGluten"`
		AnimalFlesh        bool   `json:"animalFlesh"`
		AnimalDerived      bool   `json:"animalDerived"`
		MeasurableByVolume bool   `json:"measurableByVolume"`
		Icon               string `json:"icon"`
	}

	// ValidIngredientUpdateInput represents what a user could set as input for updating valid ingredients.
	ValidIngredientUpdateInput struct {
		Name               string `json:"name"`
		Variant            string `json:"variant"`
		Description        string `json:"description"`
		Warning            string `json:"warning"`
		ContainsEgg        bool   `json:"containsEgg"`
		ContainsDairy      bool   `json:"containsDairy"`
		ContainsPeanut     bool   `json:"containsPeanut"`
		ContainsTreeNut    bool   `json:"containsTreeNut"`
		ContainsSoy        bool   `json:"containsSoy"`
		ContainsWheat      bool   `json:"containsWheat"`
		ContainsShellfish  bool   `json:"containsShellfish"`
		ContainsSesame     bool   `json:"containsSesame"`
		ContainsFish       bool   `json:"containsFish"`
		ContainsGluten     bool   `json:"containsGluten"`
		AnimalFlesh        bool   `json:"animalFlesh"`
		AnimalDerived      bool   `json:"animalDerived"`
		MeasurableByVolume bool   `json:"measurableByVolume"`
		Icon               string `json:"icon"`
	}

	// ValidIngredientDataManager describes a structure capable of storing valid ingredients permanently.
	ValidIngredientDataManager interface {
		ValidIngredientExists(ctx context.Context, validIngredientID uint64) (bool, error)
		GetValidIngredient(ctx context.Context, validIngredientID uint64) (*ValidIngredient, error)
		GetAllValidIngredientsCount(ctx context.Context) (uint64, error)
		GetAllValidIngredients(ctx context.Context, resultChannel chan []ValidIngredient) error
		GetValidIngredients(ctx context.Context, filter *QueryFilter) (*ValidIngredientList, error)
		GetValidIngredientsWithIDs(ctx context.Context, limit uint8, ids []uint64) ([]ValidIngredient, error)
		CreateValidIngredient(ctx context.Context, input *ValidIngredientCreationInput) (*ValidIngredient, error)
		UpdateValidIngredient(ctx context.Context, updated *ValidIngredient) error
		ArchiveValidIngredient(ctx context.Context, validIngredientID uint64) error
	}

	// ValidIngredientDataServer describes a structure capable of serving traffic related to valid ingredients.
	ValidIngredientDataServer interface {
		CreationInputMiddleware(next http.Handler) http.Handler
		UpdateInputMiddleware(next http.Handler) http.Handler

		SearchHandler(res http.ResponseWriter, req *http.Request)
		ListHandler(res http.ResponseWriter, req *http.Request)
		CreateHandler(res http.ResponseWriter, req *http.Request)
		ExistenceHandler(res http.ResponseWriter, req *http.Request)
		ReadHandler(res http.ResponseWriter, req *http.Request)
		UpdateHandler(res http.ResponseWriter, req *http.Request)
		ArchiveHandler(res http.ResponseWriter, req *http.Request)
	}
)

// Update merges an ValidIngredientInput with a valid ingredient.
func (x *ValidIngredient) Update(input *ValidIngredientUpdateInput) {
	if input.Name != "" && input.Name != x.Name {
		x.Name = input.Name
	}

	if input.Variant != "" && input.Variant != x.Variant {
		x.Variant = input.Variant
	}

	if input.Description != "" && input.Description != x.Description {
		x.Description = input.Description
	}

	if input.Warning != "" && input.Warning != x.Warning {
		x.Warning = input.Warning
	}

	if input.ContainsEgg != x.ContainsEgg {
		x.ContainsEgg = input.ContainsEgg
	}

	if input.ContainsDairy != x.ContainsDairy {
		x.ContainsDairy = input.ContainsDairy
	}

	if input.ContainsPeanut != x.ContainsPeanut {
		x.ContainsPeanut = input.ContainsPeanut
	}

	if input.ContainsTreeNut != x.ContainsTreeNut {
		x.ContainsTreeNut = input.ContainsTreeNut
	}

	if input.ContainsSoy != x.ContainsSoy {
		x.ContainsSoy = input.ContainsSoy
	}

	if input.ContainsWheat != x.ContainsWheat {
		x.ContainsWheat = input.ContainsWheat
	}

	if input.ContainsShellfish != x.ContainsShellfish {
		x.ContainsShellfish = input.ContainsShellfish
	}

	if input.ContainsSesame != x.ContainsSesame {
		x.ContainsSesame = input.ContainsSesame
	}

	if input.ContainsFish != x.ContainsFish {
		x.ContainsFish = input.ContainsFish
	}

	if input.ContainsGluten != x.ContainsGluten {
		x.ContainsGluten = input.ContainsGluten
	}

	if input.AnimalFlesh != x.AnimalFlesh {
		x.AnimalFlesh = input.AnimalFlesh
	}

	if input.AnimalDerived != x.AnimalDerived {
		x.AnimalDerived = input.AnimalDerived
	}

	if input.MeasurableByVolume != x.MeasurableByVolume {
		x.MeasurableByVolume = input.MeasurableByVolume
	}

	if input.Icon != "" && input.Icon != x.Icon {
		x.Icon = input.Icon
	}
}

// ToUpdateInput creates a ValidIngredientUpdateInput struct for a valid ingredient.
func (x *ValidIngredient) ToUpdateInput() *ValidIngredientUpdateInput {
	return &ValidIngredientUpdateInput{
		Name:               x.Name,
		Variant:            x.Variant,
		Description:        x.Description,
		Warning:            x.Warning,
		ContainsEgg:        x.ContainsEgg,
		ContainsDairy:      x.ContainsDairy,
		ContainsPeanut:     x.ContainsPeanut,
		ContainsTreeNut:    x.ContainsTreeNut,
		ContainsSoy:        x.ContainsSoy,
		ContainsWheat:      x.ContainsWheat,
		ContainsShellfish:  x.ContainsShellfish,
		ContainsSesame:     x.ContainsSesame,
		ContainsFish:       x.ContainsFish,
		ContainsGluten:     x.ContainsGluten,
		AnimalFlesh:        x.AnimalFlesh,
		AnimalDerived:      x.AnimalDerived,
		MeasurableByVolume: x.MeasurableByVolume,
		Icon:               x.Icon,
	}
}
